File: rust/src/rtree.rs
//use crate::triangle::{intersect, Hit, Triangle};

use crate::triangle::{intersect, Hit, Triangle};
use crate::vector::{Vector, Vector2};
use crate::ObjectLoader::load_obj;
use std::f32;
use std::rc::Rc;

#[derive(Clone, Debug)]
struct BoundingBox {
    min_point: Vector,
    max_point: Vector,
}

impl BoundingBox {
    fn intersects(&self, origin: Vector, direction: Vector) -> bool {
        fn compute_t(plane_coord: f32, origin_coord: f32, direction_coord: f32) -> f32 {
            if direction_coord == 0.0 {
                if origin_coord < plane_coord {
                    return f32::INFINITY;
                } else {
                    return f32::NEG_INFINITY;
                }
            }
            (plane_coord - origin_coord) / direction_coord
        }

        let t1 = compute_t(self.min_point.x, origin.x, direction.x);
        let t2 = compute_t(self.max_point.x, origin.x, direction.x);
        let t3 = compute_t(self.min_point.y, origin.y, direction.y);
        let t4 = compute_t(self.max_point.y, origin.y, direction.y);
        let t5 = compute_t(self.min_point.z, origin.z, direction.z);
        let t6 = compute_t(self.max_point.z, origin.z, direction.z);

        let tmin = t1.min(t2).max(t3.min(t4)).max(t5.min(t6));
        let tmax = t1.max(t2).min(t3.max(t4)).min(t5.max(t6));
        tmax > tmin.max(0.0)
    }
}

struct RTreeNode {
    bounding_box: BoundingBox,
    triangles: Vec<Rc<Triangle>>,
    children: Vec<RTreeNode>,
}

pub struct RTree {
    root: Option<RTreeNode>,
    triangles: Vec<Rc<Triangle>>,
}

impl RTree {
    pub fn new(triangles: Vec<Rc<Triangle>>, max_triangles_per_leaf: usize) -> Self {
        let root = Self::build_tree(&triangles, max_triangles_per_leaf);
        RTree { root, triangles }
    }

    fn build_tree(triangles: &[Rc<Triangle>], max_triangles_per_leaf: usize) -> Option<RTreeNode> {
        if triangles.is_empty() {
            return None;
        }

        if triangles.len() <= max_triangles_per_leaf {
            let (min_point, max_point) = Self::compute_bounds(triangles);
            let bounding_box = BoundingBox {
                min_point,
                max_point,
            };
            println!("New node!");
            return Some(RTreeNode {
                bounding_box,
                triangles: triangles.to_vec(),
                children: Vec::new(),
            });
        }

        let axis = Self::choose_split_axis(triangles);
        let mut sorted_triangles = triangles.to_vec();
        sorted_triangles.sort_by(|a, b| {
            a.coords[0].getArr()[axis as usize]
                .partial_cmp(&b.coords[0].getArr()[axis as usize])
                .unwrap()
        });

        let mid = sorted_triangles.len() / 2;
        let left_child = Self::build_tree(&sorted_triangles[..mid], max_triangles_per_leaf);
        let right_child = Self::build_tree(&sorted_triangles[mid..], max_triangles_per_leaf);

        let min_point = Vector {
            x: left_child
                .as_ref()
                .unwrap()
                .bounding_box
                .min_point
                .x
                .min(right_child.as_ref().unwrap().bounding_box.min_point.x),
            y: left_child
                .as_ref()
                .unwrap()
                .bounding_box
                .min_point
                .y
                .min(right_child.as_ref().unwrap().bounding_box.min_point.y),
            z: left_child
                .as_ref()
                .unwrap()
                .bounding_box
                .min_point
                .z
                .min(right_child.as_ref().unwrap().bounding_box.min_point.z),
        };
        let max_point = Vector {
            x: left_child
                .as_ref()
                .unwrap()
                .bounding_box
                .max_point
                .x
                .max(right_child.as_ref().unwrap().bounding_box.max_point.x),
            y: left_child
                .as_ref()
                .unwrap()
                .bounding_box
                .max_point
                .y
                .max(right_child.as_ref().unwrap().bounding_box.max_point.y),
            z: left_child
                .as_ref()
                .unwrap()
                .bounding_box
                .max_point
                .z
                .max(right_child.as_ref().unwrap().bounding_box.max_point.z),
        };
        let bounding_box = BoundingBox {
            min_point,
            max_point,
        };

        Some(RTreeNode {
            bounding_box,
            triangles: Vec::new(),
            children: vec![left_child.unwrap(), right_child.unwrap()],
        })
    }

    fn compute_bounds(triangles: &[Rc<Triangle>]) -> (Vector, Vector) {
        let mut min_point = Vector {
            x: f32::INFINITY,
            y: f32::INFINITY,
            z: f32::INFINITY,
        };
        let mut max_point = Vector {
            x: f32::NEG_INFINITY,
            y: f32::NEG_INFINITY,
            z: f32::NEG_INFINITY,
        };

        for triangle in triangles {
            for vertex in &triangle.coords {
                min_point.x = min_point.x.min(vertex.x);
                min_point.y = min_point.y.min(vertex.y);
                min_point.z = min_point.z.min(vertex.z);
                max_point.x = max_point.x.max(vertex.x);
                max_point.y = max_point.y.max(vertex.y);
                max_point.z = max_point.z.max(vertex.z);
            }
        }

        (min_point, max_point)
    }

    fn choose_split_axis(triangles: &[Rc<Triangle>]) -> usize {
        let mut extents = [0.0; 3];
        for axis in 0..3 {
            let (min_coord, max_coord): (Vec<f32>, Vec<f32>) = triangles
                .iter()
                .map(|t| (t.coords[0].getArr()[axis], t.coords[2].getArr()[axis]))
                .unzip();
            extents[axis] = max_coord.iter().cloned().fold(f32::NEG_INFINITY, f32::max)
                - min_coord.iter().cloned().fold(f32::INFINITY, f32::min);
        }
        extents
            .iter()
            .enumerate()
            .max_by(|a, b| a.1.partial_cmp(b.1).unwrap())
            .unwrap()
            .0
    }

    pub fn query_ray(&self, origin: Vector, direction: Vector) -> Vec<Rc<Triangle>> {
        self._query_ray_recursive(&self.root.as_ref().unwrap(), origin, direction)
    }

    fn _query_ray_recursive(
        &self,
        node: &RTreeNode,
        origin: Vector,
        direction: Vector,
    ) -> Vec<Rc<Triangle>> {
        if !node.bounding_box.intersects(origin, direction) {
            return vec![];
        }

        if !node.children.is_empty() {
            let mut result = Vec::new();
            for child in &node.children {
                result.extend(self._query_ray_recursive(child, origin, direction));
            }
            result
        } else {
            node.triangles.clone()
        }
    }

    pub fn check_intersections(&self, st: Vector, dir: Vector) -> Option<Hit> {
        let mut min_t = f32::INFINITY;
        let mut best_hit: Option<Hit> = None;
        let triangles = self.query_ray(st, dir);
        for tri in triangles {
            let new_hit = intersect(&Rc::clone(&tri), st, dir);
            if let Some(hit) = new_hit {
                if hit.t < min_t {
                    min_t = hit.t;
                    best_hit = Some(hit);
                }
            }
        }
        best_hit
    }
}

File: GeometricObjects/RTree.py
from __future__ import annotations
import math
from Vector import *
from Triangle import *

class BoundingBox:
    def __init__(self, min_point: list[float] | Vector, max_point: list[float] | Vector):
        """Bounding box used for R-tree construction and searching.

        Args:
            min_point (list[float] | Vector): _description_
            max_point (list[float] | Vector): _description_
        """
        self.min_point = min_point
        self.max_point = max_point

    def intersects(self, origin: Vector, direction: Vector)-> bool:
        def compute_t(plane_coord, origin_coord, direction_coord)->float:
                if direction_coord == 0:
                    if origin_coord < plane_coord:
                        return float('inf')
                    else:
                        return float('-inf')
                return (plane_coord - origin_coord) / direction_coord

        t1 = compute_t(self.min_point[0], origin[0], direction[0])
        t2 = compute_t(self.max_point[0], origin[0], direction[0])
        t3 = compute_t(self.min_point[1], origin[1], direction[1])
        t4 = compute_t(self.max_point[1], origin[1], direction[1])
        t5 = compute_t(self.min_point[2], origin[2], direction[2])
        t6 = compute_t(self.max_point[2], origin[2], direction[2])

        tmin: float = max(min(t1, t2), min(t3, t4), min(t5, t6))
        tmax: float = min(max(t1, t2), max(t3, t4), max(t5, t6))
        return tmax > max(tmin, 0)


class RTreeNode:
    def __init__(self, bounding_box: BoundingBox, triangles: list[Triangle] | None =None, children: list[RTreeNode] | None =None):
        """R-Tree node.

        Args:
            bounding_box (BoundingBox): _description_
            triangles (_type_, optional): _description_. Defaults to None.
            children (_type_, optional): _description_. Defaults to None.
        """
        self.bounding_box = bounding_box
        self.triangles = triangles if triangles is not None else []
        self.children = children if children is not None else []


class RTree:
    def __init__(self, triangles: list[Triangle], max_triangles_per_leaf: int =8):
        self.nnodes = 0
        self.root: RTreeNode | None = self.build_tree(triangles, max_triangles_per_leaf)
        self.triangles = triangles

    def build_tree(self, triangles: list[Triangle], max_triangles_per_leaf):
        if len(triangles) == 0:
            return
        #self.nnodes += 1
        #print(len(triangles), max_triangles_per_leaf)
        if len(triangles) <= max_triangles_per_leaf:
            min_point, max_point = self.compute_bounds(triangles)
            bounding_box = BoundingBox(min_point, max_point)
            #print("new node!")
            self.nnodes += 1
            return RTreeNode(bounding_box, triangles=triangles)

        # Choose an axis to split along
        axis = self.choose_split_axis(triangles)

        # Sort triangles along the chosen axis
        triangles.sort(key=lambda t: t.a.arr[axis])

        # Split triangles into two halves
        mid = len(triangles) // 2
        left_triangles = triangles[:mid]
        right_triangles = triangles[mid:]

        left_child = self.build_tree(left_triangles, max_triangles_per_leaf)
        right_child = self.build_tree(right_triangles, max_triangles_per_leaf)

        min_point = [
            min(
                left_child.bounding_box.min_point[i],
                right_child.bounding_box.min_point[i],
            )
            for i in range(3)
        ]
        max_point = [
            max(
                left_child.bounding_box.max_point[i],
                right_child.bounding_box.max_point[i],
            )
            for i in range(3)
        ]
        bounding_box = BoundingBox(min_point, max_point)

        return RTreeNode(bounding_box, children=[left_child, right_child])

    def compute_bounds(self, triangles: list[Triangle]):
        min_point = [math.inf] * 3
        max_point = [-math.inf] * 3

        for triangle in triangles:
            for vertex in [triangle.a, triangle.b, triangle.c]:
                for i in range(3):
                    min_point[i] = min(min_point[i], vertex.arr[i])
                    max_point[i] = max(max_point[i], vertex.arr[i])

        return min_point, max_point

    def choose_split_axis(self, triangles: list[Triangle]):
        extents = []
        for axis in range(3):
            min_coord, max_coord = zip(
                *[(t.a.arr[axis], t.c.arr[axis]) for t in triangles]
            )
            extents.append(max(max_coord) - min(min_coord))
        return extents.index(max(extents))

    def query_ray(self, origin: Vector, direction: Vector):
        return self._query_ray_recursive(self.root, origin, direction)

    def _query_ray_recursive(self, node: RTreeNode, origin: Vector, direction: Vector):
        if not node.bounding_box.intersects(origin, direction):
            return []

        if node.children:
            result = []
            for child in node.children:
                result.extend(self._query_ray_recursive(child, origin, direction))
            return result
        else:
            return node.triangles
        
    def checkIntersections(self, st: Vector, dir: Vector)-> Hit:
        """Checks intersections and returns the Hit object that has
        the minimum t value (first object of intersection)

        Args:
            st (Vector): Starting point of vec
            dir (Vector): Direction vector

        Returns:
            Hit : Return hit that has nones for most values, only start and dir 
            are populated
        """
        min_t = math.inf
        best_hit = Hit(False, origin=st, dir=dir) #Hit(None, None, None, None, None, st, dir)
        triangles = self.query_ray(st, dir)
        for tri in triangles:
            newHit: Hit = tri.intersect(st, dir)
            if(newHit.t and newHit.t < min_t):
                min_t = newHit.t
                best_hit = newHit
        return best_hit




if __name__ == "__main__":
    triangles = []  # List of Triangle objects
    rtree = RTree(triangles)
    origin = [0, 0, 0]
    direction = [1, 1, 1]
    intersected_triangles = rtree.query_ray(origin, direction)